src/blister.h

```diff
@@ -352,6 +357,148 @@ extern void ConvBlist (
 
 /****************************************************************************
 **
+*F  CopyBits( <fromblock>, <from-starting-bit>, <toblock>, <to-starting-bit>,
+**            <numbits> )
+**
+**  `CopyBits' copies <numbits> bits (numbering bits within a UInt
+**   from the least significant to the most significant) starting with
+**   bit number <from-starting-bit> of UInt *<fromblock> to a destination
+**   starting at bit <to-starting-bit> of *<toblock>.
+**
+**   The full function is provided as an inline function here because it is
+**   used in both blister.c and vecgf2.c and the compiler can do significant
+**   optimisations if some or all of <frombits>, <tobits> and <numbits> are
+**   known at compile time.
+*/
+
+/* constructs a mask that selects bits <from> to <to> inclusive of a UInt */
+static inline UInt MaskForCopyBits(UInt from, UInt to)
+{
+    return ((to == BIPEB - 1) ? 0 : (1L << (to + 1))) - (1L << from);
+}
+
+/* copies a block of bits from the UInt <from> to the one pointed at
+   by <to> the block starts in <from> at <startbit> and ends at
+   <endbit>. It is shifted left by <shift> and must still fit within a
+   single word (so endbits + shift must be < BIPEB and frombits +
+   shift must be non-negative */
+
+static inline void
+CopyInWord(UInt * to, UInt startbit, UInt endbit, UInt from, Int shift)
+{
+    UInt m = MaskForCopyBits(startbit + shift, endbit + shift);
+    *to &= ~m;
+    if (shift >= 0)
+        *to |= ((from << shift) & m);
+    else
+        *to |= ((from >> -shift) & m);
+}
+
+
+static ALWAYS_INLINE void CopyBits(const UInt * fromblock,
+                                   UInt         frombit,
+                                   UInt *       toblock,
+                                   UInt         tobit,
+                                   UInt         nbits)
+{
+    UInt tailbits;
+    UInt x;
+    UInt wholeblocks;
+    if (!nbits)
+        return;
+    /* Normalize */
+    fromblock += frombit / BIPEB;
+    frombit %= BIPEB;
+    toblock += tobit / BIPEB;
+    tobit %= BIPEB;
```

Comment:
This function is called from two places, and both already normalize. Is the optimizer clever enough to get rid of the second, redundant normalization?

Otherwise, I'd either remove the normalization in both calling places, or (the approach I personally favor slightly) replace the four lines above by
```C
GAP_ASSERT(frombit < BIPEB);
GAP_ASSERT(tobit < BIPEB);
```

Reply:
Fine. I put it in because it made it easier to write my C level tests, but you're right it's not needed in GAP.